### DISCUSSION TITLE
How do I integrate opencv as opencv_contrib for my JAVA project

### DISCUSSION THREAD

**Discussion by David-Braz** [community/community#143417] [Programming Help]
How do I integrate opencv as opencv_contrib, I need to add the extra modules and use them in JAVA with the Intellij IDE, if you can give me step by step I will be very grateful, I have already managed to download opencv and opencv_contrib and I have already added the modules via cmake, then I compiled it through visual studio, but I don't know what to do after that, how do I import the library into my java project?

**Comment by joegeorge022** [comment]
## Hopefully this helps....

## 1. Set Up Your OpenCV `.dll` File
- Find the `opencv_java460.dll` file (or the specific version you’re using) in your OpenCV build folder. It’s typically located in `opencv/build/bin/Release/`.
- Copy this `.dll` file to a convenient location, like `C:\opencv_java`.

## 2. Add OpenCV's Java Library (`.jar`) to IntelliJ
- Open IntelliJ and go to **File > Project Structure > Libraries**.
- Click the **+** button to add a new library, then locate and select the `opencv-460.jar` file (or your version) found in the `opencv/build/java/` folder.

## 3. Tell IntelliJ Where to Find the `.dll` File
- In IntelliJ, go to **Run > Edit Configurations**.
- Under **VM options**, add the following line (adjust the path if necessary):

  ```plaintext
  -Djava.library.path=C:\opencv_java
  ```
  
  This line tells Java where to look for the `opencv_java.dll` file you copied earlier.

## 4. Test It Out
- Create a simple Java file to verify the OpenCV setup:

  ```java
  import org.opencv.core.Core;

  public class OpenCVTest {
      static { System.loadLibrary(Core.NATIVE_LIBRARY_NAME); }
      
      public static void main(String[] args) {
          System.out.println("OpenCV version: " + Core.VERSION);
      }
  }
  ```

- Run this code. If it prints the OpenCV version, you’re all set!

## 5. Use `opencv_contrib` Modules
Since you've already included the extra modules when building, you can now use them in your Java project just like any other OpenCV feature.

---

**Reply by David-Braz** [reply]
I managed to understand the part about importing the library, but my doubt is when I compiled opencv + opencv_contrib, it generated a new jar and dlls, but when I import it into Intellij it only recognizes the Jar as a Class, I go and I import the dll as Native Library to see if it works, but it doesn't work. 

Would you have any suggestions? My main objective is to be able to integrate the modules extracted from opencv contrib to be able to use LBPHFaceRecognizer to train my facial recognition model.

Thank you for your attention.

![imagem_2024-11-04_094916461](https://github.com/user-attachments/assets/56665ce6-ecac-4403-bc19-8ee118806bbc)

**Comment by mgunawardhana** [comment]
> ### Body
> How do I integrate opencv as opencv_contrib, I need to add the extra modules and use them in JAVA with the Intellij IDE, if you can give me step by step I will be very grateful, I have already managed to download opencv and opencv_contrib and I have already added the modules via cmake, then I compiled it through visual studio, but I don't know what to do after that, how do I import the library into my java project?
> 
> ### Guidelines
> * [x]  I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.


To integrate `opencv_contrib` in your Java project using IntelliJ, after you’ve successfully built the libraries using CMake and Visual Studio, here’s a step-by-step guide to get things up and running:

### 1. Verify the Compiled Files
After building OpenCV with `opencv_contrib` modules, confirm that the compiled `.dll` (for Windows) or `.so` (for Linux/macOS) files were generated. These should be located in the `build` folder, usually under `build/bin` or `build/lib`.

### 2. Set Up Java Native Library Path
In your Java project, you'll need to tell Java where to find the OpenCV native libraries.

1. **Locate the native library**: Find the folder with the compiled `.dll` or `.so` files.
2. **Set the library path**: In IntelliJ, go to **File** > **Project Structure** > **Modules**. Then, in the **Dependencies** tab, add the path to the folder containing the native libraries as a VM option.
   - Example VM option: `-Djava.library.path="path_to_opencv_build/lib"`

### 3. Add OpenCV JAR to Your Project
1. Download the `opencv-<version>.jar` file if you haven’t done so already. This file should be in the `build/bin` directory, or you can find it from OpenCV releases online.
2. In IntelliJ, go to **File** > **Project Structure** > **Libraries**, click **+**, and add the `opencv-<version>.jar` to your project.

### 4. Load OpenCV in Your Java Code
In your Java code, load the native library by adding this line at the beginning of your program:

   ```java
   System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
   ```

   Ensure `Core.NATIVE_LIBRARY_NAME` matches the library name. For OpenCV, it’s typically `opencv_java<version>`, where `<version>` corresponds to the version number you built.

### 5. Verify Your Setup
To check if everything is working, add a simple OpenCV code snippet in your project:

   ```java
   import org.opencv.core.Core;

   public class Main {
       public static void main(String[] args) {
           System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
           System.out.println("OpenCV loaded successfully");
       }
   }
   ```

Run the code. If no errors are thrown and you see "OpenCV loaded successfully," then the setup is complete!

if your issue was fixed make this as a answer if it is not we can find more ways to that

**Reply by David-Braz** [reply]
I import the normal opencv library, but what do I do to import the opencv_contrib modules that I compiled with Visual Studio?

Thanks for your attention, it helps me a lot.

**Reply by mgunawardhana** [reply]
if your issue is fix make this as an answer if it is not we can discuss more
